schnorr_fun/src/frost.rs

```diff
@@ -661,14 +663,34 @@ impl<H: Digest<OutputSize = U32> + Clone, NG> Frost<H, NG> {
         keygen_hash.finalize().into()
     }
 
-    /// Collect all the public polynomials into a [`KeyGen`] session with a [`FrostKey`].
+    /// Collect all the public polynomials commitments into a [`KeyGen`] to produce a [`FrostKey`].
+    ///
+    /// It is crucial that at least one of these polynomials was not adversarially produced
+    /// otherwise the adversary will know the eventual secret key.
     ///
-    /// Takes a vector of point polynomials to use for this [`FrostKey`].
-    /// Also prepares a vector of verification shares for later.
-    pub fn new_keygen(
+    /// As a safety mechanism `local_secret_polys` allows you to pass in the secret scalar
+    /// polynomials you control which will be converted into the public form internally. This way
+    /// you don't trust what's in `point_polys` for the entries that you control. This protects
+    /// against a malicious adversary who publishes a `point_polys` which replaces your entries with
+    /// polynomial commitments it creates (otherwise you have to do this check yourself).
+    ///
+    /// If an entry is in both `point_polys` and `local_secret_polys` it will be silently
+    /// overwritten with the one from `local_secret_polys`.
+    pub fn new_keygen<S>(
         &self,
-        point_polys: BTreeMap<PartyIndex, Vec<Point>>,
-    ) -> Result<KeyGen, NewKeyGenError> {
+        mut point_polys: BTreeMap<PartyIndex, Vec<Point>>,
+        local_secret_polys: &BTreeMap<PartyIndex, S>,
+    ) -> Result<KeyGen, NewKeyGenError>
+    where
+        S: AsRef<[Scalar]>,
+    {
+        for (party_id, scalar_poly) in local_secret_polys {
```

Comment:
Is this futureproofing for having a single party control multiple shares of the FROST key?

Reply:
No it allows you to contribute multiple secret polynomials. Should probably put on the docs that it doesn't make sense to design a key genration protocol where you do this.